File: graci/visualizer.py
"""Immutable, observer-only Phase 5A visualizer state and event contracts.

This module projects trusted runtime facts.  It has no provider, tool, memory-store,
routing, filesystem, or network capability and therefore cannot become an authority.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, fields, is_dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Any

from .availability import Mo2State
from .registry import GLM_MODEL_ID, QWEN_MODEL_ID, HealthState, NodeRole


SNAPSHOT_SCHEMA_VERSION = 1
EVENT_SCHEMA_VERSION = 1
RECENT_EVENT_LIMIT = 100
TASK_SUMMARY_LIMIT = 240
EVENT_MESSAGE_LIMIT = 320
ERROR_SUMMARY_LIMIT = 240
DISPLAY_LABEL_LIMIT = 160
METADATA_ENTRY_LIMIT = 16
METADATA_KEY_LIMIT = 64
METADATA_VALUE_LIMIT = 160


class SystemState(str, Enum):
    IDLE = "idle"
    LISTENING = "listening"
    PLANNING = "planning"
    RETRIEVING_MEMORY = "retrieving_memory"
    REASONING = "reasoning"
    EXECUTING_TOOL = "executing_tool"
    TESTING = "testing"
    REVIEWING = "reviewing"
    ADJUDICATING = "adjudicating"
    COMPLETED = "completed"
    WARNING = "warning"
    FAILED = "failed"
    SPEAKING = "speaking"


class ActivityState(str, Enum):
    INACTIVE = "inactive"
    WAITING = "waiting"
    ACTIVE = "active"
    COMPLETED = "completed"
    FAILED = "failed"
    UNKNOWN = "unknown"


class WorkflowStatus(str, Enum):
    NOT_APPLICABLE = "not_applicable"
    UNKNOWN = "unknown"
    PENDING = "pending"
    ACTIVE = "active"
    PASSED = "passed"
    FAILED = "failed"
    REJECTED = "rejected"


class MemoryMode(str, Enum):
    NOT_APPLICABLE = "not_applicable"
    OPTIONAL = "optional"
    REQUIRED = "required"


class OperationCategory(str, Enum):
    INSPECT = "inspect"
    READ = "read"
    CREATE = "create"
    EDIT = "edit"
    COMMAND = "command"
    TEST = "test"
    VERIFY = "verify"
    REPAIR = "repair"
    ROUTE = "route"
    MEMORY_SELECT = "memory_select"


class EventSeverity(str, Enum):
    INFO = "info"
    ACTIVITY = "activity"
    SUCCESS = "success"
    WARNING = "warning"
    ERROR = "error"


class EventType(str, Enum):
    SYSTEM_READY = "system_ready"
    SYSTEM_IDLE = "system_idle"
    TASK_STARTED = "task_started"
    TASK_COMPLETED = "task_completed"
    TASK_FAILED = "task_failed"
    ROUTE_SELECTED = "route_selected"
    ROUTE_FALLBACK = "route_fallback"
    NODE_INELIGIBLE = "node_ineligible"
    MO2_RUNNING = "mo2_running"
    ENDPOINT_UNHEALTHY = "endpoint_unhealthy"
    QWEN_STARTED = "qwen_started"
    QWEN_COMPLETED = "qwen_completed"
    GLM_STARTED = "glm_started"
    GLM_COMPLETED = "glm_completed"
    MEMORY_REQUESTED = "memory_requested"
    MEMORY_SELECTED = "memory_selected"
    NO_APPLICABLE_MEMORY = "no_applicable_memory"
    MEMORY_CONFLICT = "memory_conflict"
    MEMORY_UNAVAILABLE = "memory_unavailable"
    TOOL_STARTED = "tool_started"
    TOOL_COMPLETED = "tool_completed"
    TOOL_FAILED = "tool_failed"
    TESTS_STARTED = "tests_started"
    TESTS_PASSED = "tests_passed"
    TESTS_FAILED = "tests_failed"
    REVIEW_STARTED = "review_started"
    REVIEW_COMPLETED = "review_completed"
    ADJUDICATION_COMPLETED = "adjudication_completed"


_SECRET_NAME = re.compile(r"(?:secret|token|password|passwd|api[_-]?key|private[_-]?key|credential)", re.I)


def _aware(value: datetime, name: str) -> None:
    if not isinstance(value, datetime) or value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{name} must be a timezone-aware datetime")


def _bounded(value: str | None, limit: int, name: str, *, allow_none: bool = True) -> str | None:
    if value is None and allow_none:
        return None
    if not isinstance(value, str):
        raise ValueError(f"{name} must be a string")
    cleaned = " ".join(value.split())
    if not cleaned:
        raise ValueError(f"{name} must not be empty")
    if len(cleaned) <= limit:
        return cleaned
    if limit == 1:
        return "…"
    return cleaned[:limit - 1].rstrip() + "…"


def _enum(value: Any, enum_type: type[Enum], name: str) -> None:
    if not isinstance(value, enum_type):
        raise ValueError(f"{name} must be a {enum_type.__name__}")


@dataclass(frozen=True)
class EventMetadata:
    """Bounded JSON-scalar metadata; keys that look like secret containers fail closed."""

    entries: tuple[tuple[str, str | int | float | bool | None], ...] = ()

    def __post_init__(self) -> None:
        if len(self.entries) > METADATA_ENTRY_LIMIT:
            raise ValueError("event metadata has too many entries")
        seen: set[str] = set()
        normalized = []
        for key, value in self.entries:
            key = _bounded(key, METADATA_KEY_LIMIT, "metadata key", allow_none=False)
            if _SECRET_NAME.search(key):
                raise ValueError("secret-looking metadata key is not display-safe")
            if key in seen:
                raise ValueError("duplicate metadata key")
            if type(value) not in {str, int, float, bool, type(None)}:
                raise ValueError("metadata values must be JSON scalars")
            if isinstance(value, str):
                value = _bounded(value, METADATA_VALUE_LIMIT, "metadata value", allow_none=False)
            seen.add(key)
            normalized.append((key, value))
        object.__setattr__(self, "entries", tuple(sorted(normalized)))


@dataclass(frozen=True)
class VisualizerEvent:
    event_id: str
    timestamp: datetime
    event_type: EventType
    severity: EventSeverity
    source: str
    message: str
    task_id: str | None = None
    metadata: EventMetadata = EventMetadata()
    schema_version: int = EVENT_SCHEMA_VERSION

    def __post_init__(self) -> None:
        if self.schema_version != EVENT_SCHEMA_VERSION:
            raise ValueError("unsupported event schema version")
        object.__setattr__(self, "event_id", _bounded(self.event_id, 128, "event_id", allow_none=False))
        _aware(self.timestamp, "event timestamp")
        _enum(self.event_type, EventType, "event_type")
        _enum(self.severity, EventSeverity, "severity")
        object.__setattr__(self, "source", _bounded(self.source, 80, "source", allow_none=False))
        object.__setattr__(self, "message", _bounded(self.message, EVENT_MESSAGE_LIMIT, "message", allow_none=False))
        object.__setattr__(self, "task_id", _bounded(self.task_id, 128, "task_id"))


@dataclass(frozen=True)
class RecentEventBuffer:
    events: tuple[VisualizerEvent, ...] = ()
    capacity: int = RECENT_EVENT_LIMIT

    def __post_init__(self) -> None:
        if not isinstance(self.capacity, int) or not 1 <= self.capacity <= RECENT_EVENT_LIMIT:
            raise ValueError(f"event capacity must be 1..{RECENT_EVENT_LIMIT}")
        if len(self.events) > self.capacity:
            raise ValueError("event buffer exceeds capacity")
        ids = [event.event_id for event in self.events]
        if len(ids) != len(set(ids)):
            raise ValueError("duplicate event IDs are not allowed")
        if any(self.events[index].timestamp > self.events[index + 1].timestamp
               for index in range(len(self.events) - 1)):
            raise ValueError("events must be ordered by nondecreasing timestamp")

    def append(self, event: VisualizerEvent, *, observed_at: datetime) -> "RecentEventBuffer":
        _aware(observed_at, "observed_at")
        if event.timestamp > observed_at:
            raise ValueError("future event timestamps are rejected")
        if any(existing.event_id == event.event_id for existing in self.events):
            raise ValueError("duplicate event ID")
        if self.events and event.timestamp < self.events[-1].timestamp:
            raise ValueError("out-of-order event timestamp")
        return RecentEventBuffer((self.events + (event,))[-self.capacity:], self.capacity)


@dataclass(frozen=True)
class TaskView:
    task_id: str | None = None
    goal_id: str | None = None
    summary: str | None = None
    started_at: datetime | None = None
    phase: str | None = None
    progress_current: int | None = None
    progress_total: int | None = None
    final_status: WorkflowStatus = WorkflowStatus.NOT_APPLICABLE
    failure_category: str | None = None
    failure_reason: str | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "task_id", _bounded(self.task_id, 128, "task_id"))
        object.__setattr__(self, "goal_id", _bounded(self.goal_id, 128, "goal_id"))
        object.__setattr__(self, "summary", _bounded(self.summary, TASK_SUMMARY_LIMIT, "task summary"))
        if self.started_at is not None:
            _aware(self.started_at, "task started_at")
        object.__setattr__(self, "phase", _bounded(self.phase, DISPLAY_LABEL_LIMIT, "task phase"))
        _enum(self.final_status, WorkflowStatus, "task final_status")
        object.__setattr__(self, "failure_category", _bounded(self.failure_category, 80, "failure category"))
        object.__setattr__(self, "failure_reason", _bounded(self.failure_reason, ERROR_SUMMARY_LIMIT, "failure reason"))
        if (self.progress_current is None) != (self.progress_total is None):
            raise ValueError("task progress must provide both current and total")
        if self.progress_current is not None and not (0 <= self.progress_current <= self.progress_total):
            raise ValueError("task progress is invalid")


@dataclass(frozen=True)
class ComputeNodeView:
    node_id: str
    role: NodeRole
    availability: ActivityState
    endpoint_health: HealthState
    eligible: bool | None
    mo2_state: Mo2State | None
    policy_reason: str | None
    assigned_model: str | None = None
    assigned_role: str | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "node_id", _bounded(self.node_id, 40, "node_id", allow_none=False))
        _enum(self.role, NodeRole, "node role")
        _enum(self.availability, ActivityState, "node availability")
        _enum(self.endpoint_health, HealthState, "endpoint health")
        if self.mo2_state is not None:
            _enum(self.mo2_state, Mo2State, "mo2_state")
        object.__setattr__(self, "policy_reason", _bounded(self.policy_reason, ERROR_SUMMARY_LIMIT, "policy reason"))
        object.__setattr__(self, "assigned_model", _bounded(self.assigned_model, DISPLAY_LABEL_LIMIT, "assigned model"))
        object.__setattr__(self, "assigned_role", _bounded(self.assigned_role, 80, "assigned role"))
        if self.node_id == "3090" and self.role is not NodeRole.PRIMARY:
            raise ValueError("3090 must be represented as primary")
        if self.node_id == "4090" and self.role is not NodeRole.OPTIONAL:
            raise ValueError("4090 must be represented as optional")
        if self.node_id == "4090" and self.mo2_state is Mo2State.RUNNING and self.eligible is not False:
            raise ValueError("MO2 RUNNING requires the 4090 to be ineligible")


@dataclass(frozen=True)
class ComputeView:
    primary_3090: ComputeNodeView
    optional_4090: ComputeNodeView


@dataclass(frozen=True)
class AgentView:
    model_id: str
    role: str
    state: ActivityState
    assigned_node: str | None = None
    activity: SystemState | None = None
    review_status: WorkflowStatus = WorkflowStatus.NOT_APPLICABLE
    started_at: datetime | None = None
    completed_at: datetime | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "model_id", _bounded(self.model_id, DISPLAY_LABEL_LIMIT, "model_id", allow_none=False))
        object.__setattr__(self, "role", _bounded(self.role, 80, "agent role", allow_none=False))
        _enum(self.state, ActivityState, "agent state")
        object.__setattr__(self, "assigned_node", _bounded(self.assigned_node, 40, "assigned node"))
        if self.activity is not None:
            _enum(self.activity, SystemState, "agent activity")
        _enum(self.review_status, WorkflowStatus, "review status")
        for name in ("started_at", "completed_at"):
            value = getattr(self, name)
            if value is not None:
                _aware(value, name)


@dataclass(frozen=True)
class AgentsView:
    qwen: AgentView
    glm: AgentView


@dataclass(frozen=True)
class MemoryView:
    requested: bool = False
    mode: MemoryMode = MemoryMode.NOT_APPLICABLE
    relevance_keys: tuple[str, ...] = ()
    selection_status: str | None = None
    selected_memory_ids: tuple[str, ...] = ()
    supplied_memory_ids: tuple[str, ...] = ()
    record_count: int = 0
    context_characters: int = 0
    conflict_count: int = 0
    corruption_count: int = 0

    def __post_init__(self) -> None:
        _enum(self.mode, MemoryMode, "memory mode")
        if len(self.relevance_keys) > 50 or len(self.selected_memory_ids) > 100 or len(self.supplied_memory_ids) > 100:
            raise ValueError("memory metadata exceeds contract bounds")
        for value in self.relevance_keys:
            _bounded(value, 128, "relevance key", allow_none=False)
        for value in self.selected_memory_ids + self.supplied_memory_ids:
            _bounded(value, 128, "memory ID", allow_none=False)
        object.__setattr__(self, "selection_status", _bounded(self.selection_status, 80, "selection status"))
        if any(type(value) is not int or value < 0 for value in
               (self.record_count, self.context_characters, self.conflict_count, self.corruption_count)):
            raise ValueError("memory counts must be nonnegative integers")
        if not self.requested and (self.mode is not MemoryMode.NOT_APPLICABLE or self.record_count):
            raise ValueError("unrequested memory must be not_applicable and empty")


@dataclass(frozen=True)
class OperationView:
    category: OperationCategory
    status: WorkflowStatus
    started_at: datetime | None = None
    completed_at: datetime | None = None
    target_label: str | None = None
    error_code: str | None = None
    error_reason: str | None = None

    def __post_init__(self) -> None:
        _enum(self.category, OperationCategory, "operation category")
        _enum(self.status, WorkflowStatus, "operation status")
        for name in ("started_at", "completed_at"):
            value = getattr(self, name)
            if value is not None:
                _aware(value, name)
        object.__setattr__(self, "target_label", _bounded(self.target_label, DISPLAY_LABEL_LIMIT, "target label"))
        object.__setattr__(self, "error_code", _bounded(self.error_code, 80, "error code"))
        object.__setattr__(self, "error_reason", _bounded(self.error_reason, ERROR_SUMMARY_LIMIT, "error reason"))


@dataclass(frozen=True)
class TestView:
    status: WorkflowStatus = WorkflowStatus.NOT_APPLICABLE
    command_label: str | None = None
    passed_count: int | None = None
    failed_count: int | None = None
    total_count: int | None = None
    duration_ms: int | None = None
    failure_reason: str | None = None

    def __post_init__(self) -> None:
        _enum(self.status, WorkflowStatus, "test status")
        object.__setattr__(self, "command_label", _bounded(self.command_label, DISPLAY_LABEL_LIMIT, "test command label"))
        object.__setattr__(self, "failure_reason", _bounded(self.failure_reason, ERROR_SUMMARY_LIMIT, "test failure reason"))
        for name in ("passed_count", "failed_count", "total_count", "duration_ms"):
            value = getattr(self, name)
            if value is not None and (type(value) is not int or value < 0):
                raise ValueError(f"{name} must be a nonnegative integer or null")
        if self.total_count is not None and self.passed_count is not None and self.failed_count is not None:
            if self.passed_count + self.failed_count != self.total_count:
                raise ValueError("test counts do not add to total")
        if self.failed_count and self.status is WorkflowStatus.PASSED:
            raise ValueError("failing deterministic tests cannot be represented as passed")


@dataclass(frozen=True)
class ReviewView:
    requested: bool = False
    reviewer_model_id: str | None = None
    reviewer_status: WorkflowStatus = WorkflowStatus.NOT_APPLICABLE
    structured_verdict: str | None = None
    finding_count: int = 0
    adjudication_status: WorkflowStatus = WorkflowStatus.NOT_APPLICABLE
    final_outcome: str | None = None
    reason_code: str | None = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "reviewer_model_id", _bounded(self.reviewer_model_id, DISPLAY_LABEL_LIMIT, "reviewer model"))
        _enum(self.reviewer_status, WorkflowStatus, "reviewer status")
        _enum(self.adjudication_status, WorkflowStatus, "adjudication status")
        if self.structured_verdict not in {None, "PASS", "FAIL"}:
            raise ValueError("review verdict must be PASS, FAIL, or null")
        if type(self.finding_count) is not int or not 0 <= self.finding_count <= 10:
            raise ValueError("finding count must be 0..10")
        object.__setattr__(self, "final_outcome", _bounded(self.final_outcome, 80, "final outcome"))
        object.__setattr__(self, "reason_code", _bounded(self.reason_code, 128, "adjudication reason"))
        if not self.requested and self.reviewer_status is not WorkflowStatus.NOT_APPLICABLE:
            raise ValueError("inactive reviewer must be not_applicable")


@dataclass(frozen=True)
class ExecutionView:
    operations: tuple[OperationView, ...] = ()
    tests: TestView = TestView()

    def __post_init__(self) -> None:
        if len(self.operations) > 25:
            raise ValueError("too many recent operations")


@dataclass(frozen=True)
class VisualizerSnapshot:
    snapshot_id: str
    generated_at: datetime
    system_state: SystemState
    task: TaskView
    compute: ComputeView
    agents: AgentsView
    memory: MemoryView
    execution: ExecutionView
    review: ReviewView
    recent_events: tuple[VisualizerEvent, ...] = ()
    schema_version: int = SNAPSHOT_SCHEMA_VERSION

    def __post_init__(self) -> None:
        if self.schema_version != SNAPSHOT_SCHEMA_VERSION:
            raise ValueError("unsupported snapshot schema version")
        object.__setattr__(self, "snapshot_id", _bounded(self.snapshot_id, 128, "snapshot_id", allow_none=False))
        _aware(self.generated_at, "generated_at")
        _enum(self.system_state, SystemState, "system_state")
        RecentEventBuffer(self.recent_events)
        if any(event.timestamp > self.generated_at for event in self.recent_events):
            raise ValueError("snapshot cannot contain future events")


@dataclass(frozen=True)
class TrustedRuntimeState:
    """Typed input facts owned by the runtime/controller, never by a model response."""

    controller_state: SystemState
    task: TaskView
    compute: ComputeView
    agents: AgentsView
    memory: MemoryView
    execution: ExecutionView
    review: ReviewView
    deterministic_tests_failed: bool = False
    deterministic_terminal_status: str | None = None

    def __post_init__(self) -> None:
        _enum(self.controller_state, SystemState, "controller_state")
        if type(self.deterministic_tests_failed) is not bool:
            raise ValueError("deterministic_tests_failed must be boolean")
        if self.deterministic_terminal_status not in {None, "PASS", "COMPLETED", "FAIL"}:
            raise ValueError("deterministic_terminal_status is not recognized")


def derive_system_state(source: TrustedRuntimeState) -> SystemState:
    """Derive terminal truth from controller/test facts; model claims are absent by design."""
    if source.deterministic_tests_failed or source.deterministic_terminal_status == "FAIL":
        return SystemState.FAILED
    if source.controller_state is SystemState.COMPLETED:
        if source.deterministic_terminal_status not in {"PASS", "COMPLETED"}:
            return SystemState.WARNING
        return SystemState.COMPLETED
    if source.controller_state is SystemState.FAILED:
        return SystemState.FAILED
    return source.controller_state


def project_snapshot(source: TrustedRuntimeState, *, snapshot_id: str,
                     generated_at: datetime, events: RecentEventBuffer | None = None
                     ) -> VisualizerSnapshot:
    """Pure deterministic projection.  Source objects are read but never mutated."""
    return VisualizerSnapshot(
        snapshot_id=snapshot_id, generated_at=generated_at,
        system_state=derive_system_state(source), task=source.task,
        compute=source.compute, agents=source.agents, memory=source.memory,
        execution=source.execution, review=source.review,
        recent_events=() if events is None else events.events)


def _json_value(value: Any) -> Any:
    if isinstance(value, Enum):
        return value.value
    if isinstance(value, datetime):
        _aware(value, "serialized timestamp")
        return value.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")
    if isinstance(value, EventMetadata):
        return {key: _json_value(item) for key, item in value.entries}
    if is_dataclass(value):
        return {field.name: _json_value(getattr(value, field.name)) for field in fields(value)}
    if isinstance(value, tuple):
        return [_json_value(item) for item in value]
    if type(value) in {str, int, float, bool, type(None)}:
        return value
    raise TypeError(f"unsupported visualizer serialization type: {type(value).__name__}")


def to_json_compatible(value: VisualizerSnapshot | VisualizerEvent) -> dict[str, Any]:
    return _json_value(value)


def serialize_visualizer(value: VisualizerSnapshot | VisualizerEvent) -> str:
    return json.dumps(to_json_compatible(value), ensure_ascii=False, sort_keys=True,
                      separators=(",", ":"))


def default_compute(*, primary_health: HealthState = HealthState.UNKNOWN,
                    optional_health: HealthState = HealthState.UNKNOWN,
                    optional_eligible: bool | None = None,
                    mo2_state: Mo2State | None = None,
                    optional_reason: str | None = "not_checked") -> ComputeView:
    """Construct the fixed topology from already-observed facts; performs no probe."""
    primary_availability = (ActivityState.UNKNOWN if primary_health is HealthState.UNKNOWN
                            else ActivityState.ACTIVE if primary_health is HealthState.HEALTHY
                            else ActivityState.FAILED)
    optional_availability = (ActivityState.UNKNOWN if optional_health is HealthState.UNKNOWN
                             else ActivityState.ACTIVE if optional_health is HealthState.HEALTHY
                             else ActivityState.FAILED)
    if mo2_state is Mo2State.RUNNING:
        optional_availability = ActivityState.INACTIVE
        optional_eligible = False
        optional_reason = "mo2_running"
    return ComputeView(
        ComputeNodeView("3090", NodeRole.PRIMARY, primary_availability,
                        primary_health, True if primary_health is HealthState.HEALTHY else None,
                        None, None),
        ComputeNodeView("4090", NodeRole.OPTIONAL, optional_availability,
                        optional_health, optional_eligible, mo2_state, optional_reason))


def inactive_agents() -> AgentsView:
    return AgentsView(
        AgentView(QWEN_MODEL_ID, "implementer", ActivityState.INACTIVE),
        AgentView(GLM_MODEL_ID, "reviewer", ActivityState.INACTIVE))

File: graci/visualizer_ui/visualizer.js
"use strict";
(() => {
  const API = "/graci/visualizer/v1";
  const SNAPSHOT_INTERVAL_MS = 3000;
  const MAX_EVENT_ROWS = 100;
  const PRESENCE_BY_STATE = Object.freeze({
    "idle":"resting", "listening":"receptive", "planning":"thinking",
    "retrieving_memory":"thinking", "reasoning":"thinking", "executing_tool":"acting",
    "testing":"validating", "reviewing":"validating", "adjudicating":"validating",
    "completed":"success", "warning":"warning", "failed":"failure", "speaking":"responding"
  });
  const SAFE_PRESENCE = "warning";
  const presenceFor = (systemState) => PRESENCE_BY_STATE[systemState] || SAFE_PRESENCE;
  const ALL_STATES = Object.freeze(Object.keys(PRESENCE_BY_STATE));
  const ALL_SEVERITIES = ["info","activity","success","warning","error"];
  const $ = (id) => document.getElementById(id);
  const state = { snapshot: null, events: new Map(), lastSuccess: 0, eventSource: null };
  const text = (value, fallback = "—") => value === null || value === undefined || value === "" ? fallback : String(value).replaceAll("_", " ").toUpperCase();
  const localTime = (value, seconds = true) => { if (!value) return "—"; const date = new Date(value); return Number.isNaN(date.valueOf()) ? "—" : date.toLocaleTimeString([], {hour:"2-digit",minute:"2-digit",second:seconds?"2-digit":undefined}); };
  const shortId = (value) => value ? String(value).slice(0, 14) : "—";
  const field = (root, name, value) => { const node = root.querySelector(`[data-field="${name}"]`); if (node) node.textContent = text(value); };
  function connection(kind, label) { document.body.dataset.connection = kind; const node=$("connection"); node.className=`connection ${kind}`; node.innerHTML=`<i></i> ${label}`; document.body.classList.toggle("stale", kind === "disconnected"); }
  function renderNode(id, node) {
    const root=$(id); if (!node) return; const blocked=node.node_id === "4090" && (node.eligible === false || node.mo2_state === "running");
    const available=node.endpoint_health === "healthy" && !blocked; root.dataset.status=blocked?"blocked":available?"available":node.endpoint_health === "unhealthy"?"failed":"unknown";
    root.querySelector(".availability").textContent=blocked?"BLOCKED — MO2 RUNNING":available?"AVAILABLE / HEALTHY":text(node.availability);
    field(root,"health",node.endpoint_health); field(root,"model",node.assigned_model,"UNASSIGNED"); field(root,"role",node.assigned_role || node.role); field(root,"mo2",node.mo2_state,"NOT OBSERVED"); field(root,"eligibility",node.eligible===true?"ELIGIBLE":node.eligible===false?"BLOCKED":"UNKNOWN"); field(root,"reason",node.policy_reason,"NO POLICY RESTRICTION");
  }
  function renderAgent(id, agent) { const root=$(id); if(!agent)return; field(root,"model",agent.model_id); field(root,"activity",agent.activity || agent.review_status || agent.state); root.dataset.status=agent.state; }
  function setStage(name,status,label) { const root=document.querySelector(`[data-stage="${name}"]`); if(!root)return; root.dataset.status=status; root.querySelector("b").textContent=text(label || status); }
  function renderPipeline(s) {
    const active=s.system_state; const memory=s.memory.requested ? (active==="retrieving_memory"?"active":s.memory.supplied_memory_ids.length?"passed":"pending") : "not_applicable";
    const ops=s.execution.operations||[]; const opFailed=ops.some(x=>x.status==="failed"); const opActive=ops.some(x=>x.status==="active");
    setStage("memory",memory); setStage("qwen",s.agents.qwen.state,s.agents.qwen.activity||s.agents.qwen.state); setStage("tools",opFailed?"failed":opActive||active==="executing_tool"?"active":ops.length?"passed":"not_applicable"); setStage("tests",s.execution.tests.status); setStage("review",s.review.reviewer_status); setStage("adjudication",s.review.adjudication_status);
  }
  function renderSnapshot(s) {
    state.snapshot=s; state.lastSuccess=Date.now(); connection("live","LIVE"); const presence=presenceFor(s.system_state); document.body.dataset.systemState=ALL_STATES.includes(s.system_state)?s.system_state:"warning"; document.body.dataset.presence=presence;
    $("overall-state").textContent=text(s.system_state,"UNKNOWN"); $("core-state").textContent=text(s.system_state,"UNKNOWN"); $("presence-category").textContent=text(presence); $("schema").textContent=`v${s.schema_version}`; $("snapshot-id").textContent=shortId(s.snapshot_id); $("updated").textContent=localTime(s.generated_at); $("freshness").textContent="LIVE OBSERVED STATE";
    renderNode("node-3090",s.compute.primary_3090); renderNode("node-4090",s.compute.optional_4090); renderAgent("agent-qwen",s.agents.qwen); renderAgent("agent-glm",s.agents.glm);
    const task=s.task, hasTask=Boolean(task.task_id); $("task-summary").textContent=hasTask?(task.summary||"BOUNDED TASK"):"SYSTEM AT REST"; $("task-id").textContent=shortId(task.task_id); $("task-phase").textContent=text(task.phase,hasTask?text(s.system_state):"IDLE"); $("task-started").textContent=localTime(task.started_at);
    const pct=task.progress_total?Math.min(100,Math.round(task.progress_current/task.progress_total*100)):0; $("task-progress").style.width=`${pct}%`; $("task-status").textContent=task.failure_reason|| (hasTask?`${text(task.final_status)} · ${pct}% observed progress`:"No current task. Monitoring local runtime.");
    const m=s.memory; document.querySelector(".memory-panel").classList.toggle("active",m.requested); $("memory-status").textContent=text(m.selection_status,m.requested?"REQUESTED":"N/A"); $("memory-mode").textContent=text(m.mode); $("memory-keys").textContent=m.relevance_keys.length; $("memory-selected").textContent=m.selected_memory_ids.length; $("memory-supplied").textContent=m.supplied_memory_ids.length; $("memory-context").textContent=Number(m.context_characters).toLocaleString(); $("memory-diag").textContent=`${m.conflict_count} / ${m.corruption_count}`;
    const review=$("review-panel"), adjudication=$("adjudication-panel"); review.querySelector("strong").textContent=text(s.review.structured_verdict||s.review.reviewer_status); review.dataset.status=s.review.reviewer_status; adjudication.querySelector("strong").textContent=text(s.review.final_outcome||s.review.adjudication_status); adjudication.dataset.status=s.review.adjudication_status;
    renderPipeline(s); renderOperations(s.execution.operations||[]); (s.recent_events||[]).forEach(addEvent); renderEvents();
  }
  function renderOperations(ops) { const root=$("operations"); const rows=ops.slice(-12).reverse(); root.innerHTML=rows.length?rows.map(op=>`<div class="operation" data-status="${op.status}"><b>${text(op.category)}</b><span>${escapeHtml(op.target_label||"BOUNDED OPERATION")}</span><em>${text(op.status)}</em></div>`).join(""):'<p class="empty">No execution operations observed.</p>'; }
  function escapeHtml(value) { const e=document.createElement("span"); e.textContent=String(value); return e.innerHTML; }
  function addEvent(event) { if(!event||!event.event_id||!ALL_SEVERITIES.includes(event.severity)||state.events.has(event.event_id))return; state.events.set(event.event_id,event); while(state.events.size>MAX_EVENT_ROWS)state.events.delete(state.events.keys().next().value); }
  function renderEvents() { const values=[...state.events.values()].slice(-MAX_EVENT_ROWS).reverse(); $("event-count").textContent=`${values.length} / ${MAX_EVENT_ROWS}`; $("events").innerHTML=values.length?values.map(e=>`<div class="event" data-severity="${e.severity}"><time title="${escapeHtml(e.timestamp)}">${localTime(e.timestamp)}</time><b>${text(e.event_type)}</b><small>${text(e.source)}</small><span>${escapeHtml(e.message)}</span></div>`).join(""):'<p class="empty">Waiting for trusted local events…</p>'; }
  async function getJson(path) { const response=await fetch(path,{cache:"no-store",headers:{"Accept":"application/json"}}); if(!response.ok)throw new Error(`HTTP ${response.status}`); return response.json(); }
  async function refreshSnapshot() { try { await getJson(`${API}/health`); renderSnapshot(await getJson(`${API}/snapshot`)); } catch { markDisconnected(); } }
  async function refreshEvents() { try { (await getJson(`${API}/events`)).forEach(addEvent); renderEvents(); } catch { markDisconnected(); } }
  function markDisconnected() { connection("disconnected","DISCONNECTED"); if(state.snapshot){const age=Math.max(0,Math.floor((Date.now()-state.lastSuccess)/1000)); $("freshness").textContent=`STALE · ${age}s SINCE UPDATE`;} }
  function connectEvents() { if(state.eventSource)state.eventSource.close(); const source=new EventSource(`${API}/events/stream`); state.eventSource=source; source.onopen=()=>{connection("live","LIVE"); refreshSnapshot();}; source.onerror=()=>markDisconnected(); ALL_STATES.forEach(()=>{}); source.onmessage=consume; ["system_ready","system_idle","task_started","task_completed","task_failed","route_selected","route_fallback","node_ineligible","mo2_running","endpoint_unhealthy","qwen_started","qwen_completed","glm_started","glm_completed","memory_requested","memory_selected","no_applicable_memory","memory_conflict","memory_unavailable","tool_started","tool_completed","tool_failed","tests_started","tests_passed","tests_failed","review_started","review_completed","adjudication_completed"].forEach(type=>source.addEventListener(type,consume)); }
  function consume(message) { try { const event=JSON.parse(message.data); addEvent(event); renderEvents(); if(["task_started","task_completed","task_failed","mo2_running","tests_passed","tests_failed","review_completed","adjudication_completed"].includes(event.event_type))refreshSnapshot(); } catch { /* malformed observer data is ignored */ } }
  function tick() { $("clock").textContent=new Date().toLocaleTimeString([], {hour:"2-digit",minute:"2-digit",second:"2-digit"}); if(document.body.dataset.connection==="disconnected"&&state.lastSuccess)markDisconnected(); }
  async function start() { tick(); setInterval(tick,1000); await Promise.allSettled([refreshSnapshot(),refreshEvents()]); connectEvents(); setInterval(refreshSnapshot,SNAPSHOT_INTERVAL_MS); }
  start();
})();
